packages/synapse-sdk/src/storage/context.ts

```diff
@@ -1353,14 +1355,73 @@ export class StorageContext {
   }
 
   /**
-   * Get the list of piece CIDs for this service service's data set by querying the PDP server.
+   * Get the list of piece CIDs for this service service's data set.
+   * Gets data directly from PDPVerifier contract (source of truth) rather than Curio.
    * @returns Array of piece CIDs as PieceCID objects
+   * @deprecated Use getPieces() generator for better memory efficiency with large data sets
    */
   async getDataSetPieces(): Promise<PieceCID[]> {
-    const dataSetData = await this._pdpServer.getDataSet(this._dataSetId)
-    return dataSetData.pieces.map((piece) => piece.pieceCid)
+    const pieces: PieceCID[] = []
+    for await (const [pieceCid] of this.getPieces()) {
+      pieces.push(pieceCid)
+    }
+    return pieces
   }
 
+  /**
+   * Get all active pieces for this data set as an async generator.
+   * This provides lazy evaluation and better memory efficiency for large data sets.
+   * Gets data directly from PDPVerifier contract (source of truth) rather than Curio.
+   * @param options - Optional configuration object
+   * @param options.batchSize - The batch size for each pagination call (default: 100)
+   * @param options.signal - Optional AbortSignal to cancel the operation
+   * @yields Tuple of [PieceCID, pieceId] - the piece ID is needed for certain operations like deletion
+   */
+  async *getPieces(options?: { batchSize?: number; signal?: AbortSignal }): AsyncGenerator<[PieceCID, number]> {
+    const pdpVerifierAddress = this._warmStorageService.getPDPVerifierAddress()
+    const pdpVerifier = new PDPVerifier(this._synapse.getProvider(), pdpVerifierAddress)
+
+    const batchSize = options?.batchSize ?? 100
+    const signal = options?.signal
+    let offset = 0
```

Comment:
we could probably accept offset as input if someone wants to continue a previous operation.. and might want to return offset as well?

Probably could use some `PaginatedOptionsInput` and `PaginatedOptionsOutput` interfaces to help with @hugomrdias's goal of normalizing generators so we don't keep redefining things anywhere we implement a generator for a paginated call